verilog/tools/ls/autoexpand.cc

```diff
@@ -89,33 +154,156 @@ class AutoExpander {
+  std::string indent_;
 
   // Regex for finding AUTOARG comments
-  static std::regex autoarg_re_;
+  static const std::regex autoarg_re_;
+
+  // Regex for finding AUTOINST comments
+  static const std::regex autoinst_re_;
+
+  // Regexes for AUTO_TEMPLATE comments
+  static const std::regex autotemplate_detect_re_;
+  static const std::regex autotemplate_type_re_;
+  static const std::regex autotemplate_conn_re_;
 };
 
-std::regex AutoExpander::autoarg_re_{R"(/\*\s*AUTOARG\s*\*/)"};
+const std::regex AutoExpander::autoarg_re_{R"(/\*\s*AUTOARG\s*\*/)"};
+
+const std::regex AutoExpander::autoinst_re_{R"(/\*\s*AUTOINST\s*\*/)"};
 
-void AutoExpander::EmitPortDeclarations(
+// AUTO_TEMPLATE regex breakdown:
+// /\*                             – start of comment
+// (\s*\S+\s+AUTO_TEMPLATE\s*\n)*  – optional other AUTO_TEMPLATE types, end
+//                                   with newline
+// \s*\S+\s+AUTO_TEMPLATE          – at least one AUTO_TEMPLATE is required
+// \s*(".*")?                      – optional instance name regex
+// \s*\([\s\S]*?\);                – parens with port connections
+// \s*\*/                          – end of comment
+const std::regex AutoExpander::autotemplate_detect_re_{
+    R"(/\*(\s*\S+\s+AUTO_TEMPLATE\s*\n)*\s*\S+\s+AUTO_TEMPLATE\s*(".*")?\s*\([\s\S]*?\);\s*\*/)"};
+
+// AUTO_TEMPLATE type regex: the first capturing group is the instance type
+const std::regex AutoExpander::autotemplate_type_re_{
+    R"((\S+)\s+AUTO_TEMPLATE)"};
+
+// AUTO_TEMPLATE connection regex breakdown:
+// \.\s*      – starts with a dot
+// ([^\s(]+?) – first group, at least one character other than whitespace or
+//              opening paren
+// \s*\(\s*   – optional whitespace, opening paren, optional whitespace again
+// ([^\s(]+?) – second group, same as the first one
+// \s*\)*     – optional whitespace, closing paren
+const std::regex AutoExpander::autotemplate_conn_re_{
+    R"(\.\s*([^\s(]+?)\s*\(\s*([^\s(]+?)\s*\))"};
+
+const AutoExpander::Template *AutoExpander::Template::from(
+    const Map &templates, const absl::string_view type_id,
+    const long int instance_offset) {
+  const auto templates_it = templates.find(type_id);
+  const Template *matching_tmpl = nullptr;
+  if (templates_it != templates.end()) {
+    // Linear search for the matching template (there should be very few
+    // templates per type, often just one)
+    for (const auto &tmpl : templates_it->second) {
+      if (instance_offset < tmpl.offset) break;
+      matching_tmpl = &tmpl;
+    }
+  }
+  return matching_tmpl;
+}
+
+void AutoExpander::Module::EmitPortDeclarations(
+    std::ostream &output, const absl::string_view indent,
+    const absl::string_view header,
+    const std::function<bool(const Port &)> &pred) const {
+  bool first = true;
+  for (const Port &port : ports_) {
+    if (!pred(port)) continue;
+    if (first) {
+      if (output.tellp() != 0) output << ',';
+      output << '\n' << indent << "// " << header << '\n' << indent;
```

Comment:
Looks like you try to properly match the formatting style.

It might be worthwhile (maybe later) to add a unit test that pulls in the formatter from other parts of Verible and runs the resulting template code through it to verify that input and output is the same. Low priority at this point I think, so in this PR adding a TODO might be sufficient.

Reply:
I get the indentation from the formatting style. It would actually be better to get it from the file itself, but I'm not sure it's worth it.